Sources/Moya/Moya+Alamofire.swift

```diff
@@ -106,7 +106,7 @@ extension DownloadRequest: Requestable {
     }
 }
 
-final class MoyaRequestInterceptor: RequestInterceptor {
+final class MoyaRequestInterceptor: NSRecursiveLock, RequestInterceptor {
```

Comment:
oh nice, first time seeing inheritance from `NSRecursiveLock` rather then using an inner lock inside the object - are there any pros/cons using it that way?

Reply:
This was only for simplicity, the concept is making the interceptor lockable, so we could manage lock/unlock inside and outside the class. The other option was to have a _lock property ... but then we would have to create an encapsulation logic to the willSend getter and setter. 
```swift
private _willSend: ...

var willSend : ... {
    get {
       _lock.lock(); defer { _lock.unlock() }
       return _willSend
    }

     set {
        _lock.lock(); defer { _lock.unlock() }
        _willSend = newValue
     }
}

```
I think we it was a public API that we wanted to make sure the access is atomic it would be a better option,because we encapsulate everything and users will not have to worry about. But since is an internal thing I think make the interceptor lockable is a more clear and simple option.
In the end the pro is just simplicity :)

Reply:
This is a very good explanation, thanks! 

I think that making the class self-contained could help us in the future (meaning we lock it all in the interceptor). When we end up needing to use the interceptor again, we won't have to think about locking from the outside because it will be already taken-care-of. What do you think?

Reply:
I personally prefer the current approach, because not all access need to be locked and most of the times the caller is the one who knows if we need locking or not.
But in this case think there is no problem in doing that, and is actually nicer at the API level (for the reason you mention that we don't to think about locking from the outside because it will be already taken-care-of).
I'll change that as soon as possible then :) 
Thanks

Reply:
Done :)